shopinvader_auth_jwt/components/service_context_provider.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2021 ACSONE SA/NV
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+import logging
+
+from odoo import _
+from odoo.exceptions import MissingError
+
+from odoo.addons.component.core import Component
+
+_logger = logging.getLogger(__name__)
+
+
+class AuthApiKeyServiceContextProvider(Component):
+    _name = "auth_jwt.shopinvader.service.context.provider"
+    _inherit = "shopinvader.service.context.provider"
+    _usage = "auth_jwt_component_context_provider"
+
+    def _get_partner(self):
+        partner_model = self.env["shopinvader.partner"]
+        partner_email = self.request.jwt_payload.get("email")
+        backend = self._get_backend()
+        if partner_email:
+            shopinvader_partner = self._find_partner(backend, partner_email)
+            if len(shopinvader_partner) == 1:
+                self._validate_partner(backend, shopinvader_partner)
+                return shopinvader_partner
+            else:
+                _logger.warning("Wrong email, jwt payload ignored")
+                if len(shopinvader_partner) > 1:
+                    _logger.warning(
+                        "More than one shopinvader.partner found for:"
+                        " backend_id={} email={}".format(backend.id, partner_email)
+                    )
+                # Could be because the email is not related to a partner or
+                # because the partner is inactive
+                raise MissingError(_("The given partner is not found!"))
+        return partner_model.browse([])
+
+    def _get_backend(self):
+        backend_model = self.env["shopinvader.backend"]
+        return backend_model._get_from_jwt_aud(self.request.jwt_payload.get("aud"))
```

Comment:
Same here, what if jwt_payload is None ? I guess it cannot happen, has there are no public /shoinvader routes ?

Reply:
Good catch.... thinking further: At this stage, with shopinvader_auth_jwt, it's not possible to create a cart while the user is not authenticated :thinking: Moreover, how to determine the backend for public requests? (From the host or origin http header?)

Reply:
It always bothered me that we determine the backend from low level authentication details. I don't think the [Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) header will help.

Perhaps we can simply use an explicit, optional http header or query parameter to address a specific backend (to discriminate when there is more than one backend in the Odoo instance) ?

We'll also need a public route to add to cart and query an anonymous cart by uuid.

Reply:
In any case in the frontend code we'll need a global connection object (a simple wrapper around js `fetch` that manages authentication), so we can add such an http header to its constructor.

Reply:
I'll add this default optional header as default parameter for all shopinvader services and manage-it into the default service context provider.